### PR TITLE
Produce an error when one leaves out keyword static in an implementation of IWSAM

### PR DESCRIPTION
**Pull request by edgarfgp**
Fixes https://github.com/dotnet/fsharp/issues/16156 and https://github.com/dotnet/fsharp/issues/14316

### CODE REVIEW

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 1]
```diff
@@ -4457,6 +4457,44 @@ module TcDeclarations =
...
+                                    |> List.filter(fun x -> not x.IsInstanceMember)
+                                Some staticAbstractMembers
+                            | _ -> None
+                    )
+                    |> List.concat
+                        
+                let implMembers =
+                    members
+                    |> List.choose(
+                        fun synMemberDef ->
```
Comment (smoothdeveloper) [inline src/Compiler/Checking/CheckDeclarations.fs:4478]: Wondering if those aren't simpler expressed as a nested for loop in a list expression; could also collapse the match and fun into function if it remains a pipe.

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 2]
```diff
@@ -4418,6 +4418,33 @@ module TcDeclarations =
...
+                | SynMemberDefn.Interface(members= Some(synMemberDefns)) ->
+                    [ for synMemberDefn in synMemberDefns do
+                        match synMemberDefn with
+                        | SynMemberDefn.Member(memberDefn = SynBinding(headPat = SynPat.LongIdent(longDotId = SynLongIdent(id = [ _; ident])))) -> ident
+                        | _ -> () ]
+                | _ -> ()
+            ]|> List.concat
+            
+        for memberId in implMembers do
+            if staticMembers |> List.exists(fun name -> name = memberId.idText)  then
```
Comment (auduchinok) [inline src/Compiler/Checking/CheckDeclarations.fs:4445]: Is it safe to compare members based just on their name? Do we want to check inferred signatures as well?

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 3]
```diff
@@ -4418,6 +4418,33 @@ module TcDeclarations =
             let isAtOriginalTyconDefn = true
             let core = MutRecDefnsPhase1DataForTycon(synTyconInfo, SynTypeDefnSimpleRepr.Exception r, implements1, false, false, isAtOriginalTyconDefn)
             core, extra_vals_Inherits_Abstractslots @ extraMembers
+
```
[the remark this answers is not in the document]
Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs:4421]: Thanks for the review. I'm looking at the [IWSAM Implementation](https://github.com/dotnet/fsharp/pull/13119/files) to see if there is a better place to do this where I have access to the implemented members and the interface definition. Maybe @vzarytovskii knows the right place?

[the remark this answers is not in the document]
Comment (smoothdeveloper) [reply src/Compiler/Checking/CheckDeclarations.fs:4421]: This is the place where "no implementation was given for" messages are being generated:

https://github.com/dotnet/fsharp/blob/72ee6e448e0c86865915580abe37df363b980f45/src/Compiler/Checking/MethodOverrides.fs#L318

it would feel like the right place, to have the logic there.

Sorry that I initially pointed you in the chat at CheckDeclaration as entry point where the slots start to be processed.

File: src/Compiler/Checking/CheckDeclarations.fs (hunk 2 above)
Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs:4445]: I think it is better to check for the inferred signatures too here.

Comment (auduchinok) [reply src/Compiler/Checking/CheckDeclarations.fs:4445]: Could you add a test that would show a difference, if overloads are present?

File: src/Compiler/Checking/CheckDeclarations.fs (hunk 3 above)
[the remark this answers is not in the document]
Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs:4421]: Moved the check to a more appropriate place now

**Comment by vzarytovskii** [comment]
Fix can be a simpler fix, I think, needs a bunch of testing though, both positive and negative cases (like the case in the linked issues, as well a the opposite - slot is instance, but override is static):

```diff
diff --git a/src/Compiler/Checking/CheckExpressions.fs b/src/Compiler/Checking/CheckExpressions.fs
index d232a04e6..11162f71c 100644
--- a/src/Compiler/Checking/CheckExpressions.fs
+++ b/src/Compiler/Checking/CheckExpressions.fs
@@ -7012,7 +7012,7 @@ and TcObjectExpr (cenv: cenv) env tpenv (objTy, realObjTy, argopt, binds, extraI
               let overrides' =
                   [ for overrideMeth in overrides do
                         let overrideInfo, (_, thisVal, methodVars, bindingAttribs, bindingBody) = overrideMeth
-                        let (Override(_, _, id, mtps, _, _, _, isFakeEventProperty, _)) = overrideInfo
+                        let (Override(_, _, id, mtps, _, _, _, isFakeEventProperty, _, _)) = overrideInfo
 
                         if not isFakeEventProperty then
                             let searchForOverride =
diff --git a/src/Compiler/Checking/MethodOverrides.fs b/src/Compiler/Checking/MethodOverrides.fs
index bfa11de4c..221921ac6 100644
--- a/src/Compiler/Checking/MethodOverrides.fs
+++ b/src/Compiler/Checking/MethodOverrides.fs
@@ -44,7 +44,8 @@ type OverrideInfo =
         argTypes: TType list list *
         returnType: TType option *
         isFakeEventProperty: bool *
-        isCompilerGenerated: bool
+        isCompilerGenerated: bool *
+        isInstance: bool
 
     member x.CanImplement = let (Override(canImplement=a)) = x in a
 
@@ -61,6 +62,8 @@ type OverrideInfo =
     member x.ReturnType = let (Override(returnType=b)) = x in b
 
     member x.IsCompilerGenerated = let (Override(isCompilerGenerated=b)) = x in b
+    
+    member x.IsInstance = let (Override(isInstance=b)) = x in b
 
 type RequiredSlot = 
     | RequiredSlot of methodInfo: MethInfo * isOptional: bool
@@ -104,7 +107,7 @@ exception OverrideDoesntOverride of DisplayEnv * OverrideInfo * MethInfo option
 module DispatchSlotChecking =
 
     /// Print the signature of an override to a buffer as part of an error message
-    let PrintOverrideToBuffer denv os (Override(_, _, id, methTypars, memberToParentInst, argTys, retTy, _, _)) = 
+    let PrintOverrideToBuffer denv os (Override(_, _, id, methTypars, memberToParentInst, argTys, retTy, _, _, _)) = 
        let denv = { denv with showTyparBinding = true }
        let retTy = (retTy  |> GetFSharpViewOfReturnType denv.g)
        let argInfos = 
@@ -136,7 +139,7 @@ module DispatchSlotChecking =
         let (CompiledSig (argTys, retTy, fmethTypars, ttpinst)) = CompiledSigOfMeth g amap m minfo
 
         let isFakeEventProperty = minfo.IsFSharpEventPropertyMethod
-        Override(parentType, minfo.ApparentEnclosingTyconRef, mkSynId m nm, fmethTypars, ttpinst, argTys, retTy, isFakeEventProperty, false)
+        Override(parentType, minfo.ApparentEnclosingTyconRef, mkSynId m nm, fmethTypars, ttpinst, argTys, retTy, isFakeEventProperty, false, minfo.IsInstance)
 
     /// Get the override info for a value being used to implement a dispatch slot.
     let GetTypeMemberOverrideInfo g reqdTy (overrideBy: ValRef) = 
@@ -175,7 +178,7 @@ module DispatchSlotChecking =
                 //CanImplementAnySlot  <<----- Change to this to enable implicit interface implementation
 
         let isFakeEventProperty = overrideBy.IsFSharpEventProperty(g)
-        Override(implKind, overrideBy.MemberApparentEntity, mkSynId overrideBy.Range nm, memberMethodTypars, memberToParentInst, argTys, retTy, isFakeEventProperty, overrideBy.IsCompilerGenerated)
+        Override(implKind, overrideBy.MemberApparentEntity, mkSynId overrideBy.Range nm, memberMethodTypars, memberToParentInst, argTys, retTy, isFakeEventProperty, overrideBy.IsCompilerGenerated, overrideBy.IsInstanceMember)
 
     /// Get the override information for an object expression method being used to implement dispatch slots
     let GetObjectExprOverrideInfo g amap (implTy, id: Ident, memberFlags, ty, arityInfo, bindingAttribs, rhsExpr) = 
@@ -200,7 +203,7 @@ module DispatchSlotChecking =
                     CanImplementAnyClassHierarchySlot
                     //CanImplementAnySlot  <<----- Change to this to enable implicit interface implementation
             let isFakeEventProperty = CompileAsEvent g bindingAttribs
-            let overrideByInfo = Override(implKind, tcrefOfAppTy g implTy, id, tps, [], argTys, retTy, isFakeEventProperty, false)
+            let overrideByInfo = Override(implKind, tcrefOfAppTy g implTy, id, tps, [], argTys, retTy, isFakeEventProperty, false, memberFlags.IsInstance)
             overrideByInfo, (baseValOpt, thisv, vs, bindingAttribs, rhsExpr)
         | _ -> 
             error(InternalError("Unexpected shape for object expression override", id.idRange))
@@ -227,7 +230,7 @@ module DispatchSlotChecking =
         
     /// Check if an override is a partial match for the requirements for a dispatch slot except for the name.
     let IsSigPartialMatch g (dispatchSlot: MethInfo) compiledSig overrideBy =
-        let (Override(_, _, _, methTypars, _, argTys, _retTy, _, _)) = overrideBy
+        let (Override(_, _, _, methTypars, _, argTys, _retTy, _, _, _)) = overrideBy
         let (CompiledSig (vargTys, _, fvmethTypars, _)) = compiledSig
         methTypars.Length = fvmethTypars.Length &&
         IsTyparKindMatch compiledSig overrideBy && 
@@ -249,7 +252,7 @@ module DispatchSlotChecking =
      
     /// Check if an override exactly matches the requirements for a dispatch slot except for the name.
     let IsSigExactMatch g amap m dispatchSlot overrideBy =
-        let (Override(_, _, _, methTypars, memberToParentInst, argTys, retTy, _, _)) = overrideBy
+        let (Override(_, _, _, methTypars, memberToParentInst, argTys, retTy, _, _, _)) = overrideBy
         let compiledSig = CompiledSigOfMeth g amap m dispatchSlot
         IsSigPartialMatch g dispatchSlot compiledSig overrideBy &&
         let (CompiledSig (vargTys, vrty, fvmethTypars, ttpinst)) = compiledSig
@@ -300,7 +303,8 @@ module DispatchSlotChecking =
     /// Check if an override exactly matches the requirements for a dispatch slot.
     let IsExactMatch g amap m dispatchSlot overrideBy =
         IsNameMatch dispatchSlot overrideBy &&
-        IsSigExactMatch g amap m dispatchSlot overrideBy
+        IsSigExactMatch g amap m dispatchSlot overrideBy &&
+        dispatchSlot.IsInstance = overrideBy.IsInstance
 
     /// Check if an override implements a dispatch slot 
     let OverrideImplementsDispatchSlot g amap m dispatchSlot availPriorOverride =
@@ -386,7 +390,7 @@ module DispatchSlotChecking =
                             noimpl()
                         | [ overrideBy ] ->
 
-                            let (Override(_, _, _, methTypars, _, argTys, _, _, _)) = overrideBy
+                            let (Override(_, _, _, methTypars, _, argTys, _, _, _, _)) = overrideBy
 
                             let moreThanOnePossibleDispatchSlot =
                                 dispatchSlots
@@ -965,4 +969,3 @@ let GetAbstractPropInfosForSynPropertyDecl(infoReader: InfoReader, ad, memberNam
         
     let dispatchSlots = pinfos |> List.filter (fun pinfo -> pinfo.IsVirtualProperty)
     dispatchSlots
-
diff --git a/src/Compiler/Checking/MethodOverrides.fsi b/src/Compiler/Checking/MethodOverrides.fsi
index d18a8cdc6..ac72aa0ca 100644
--- a/src/Compiler/Checking/MethodOverrides.fsi
+++ b/src/Compiler/Checking/MethodOverrides.fsi
@@ -32,7 +32,8 @@ type OverrideInfo =
         argTypes: TType list list *
         returnType: TType option *
         isFakeEventProperty: bool *
-        isCompilerGenerated: bool
+        isCompilerGenerated: bool *
+        isInstance: bool
 
     member ArgTypes: TType list list
 
@@ -50,6 +51,8 @@ type OverrideInfo =
 
     member ReturnType: TType option
 
+    member IsInstance: bool
+
 type RequiredSlot =
     | RequiredSlot of methodInfo: MethInfo * isOptional: bool
     | DefaultInterfaceImplementationSlot of methodInfo: MethInfo * isOptional: bool * possiblyNoMostSpecific: bool

```

**Comment by vzarytovskii** [comment]
Maybe also verifying that tooling (service tests) light up the issue too. And maybe do different comparison between two bools, to make sure we don't box.

**Comment by edgarfgp** [comment]
Thanks for the detailed analysis, @vzarytovskii. Seems easier fix in my last commit(Lets see what the CI says). I will have  look at your implementation suggestion after :)

**Comment by vzarytovskii** [comment]
> Thanks for the detailed analysis, @vzarytovskii. Seems easier fix in my last commit(Lets see what the CI says). 

It looks like it adds a bunch more. Change in `IsExactMatch` will also cover object expressions for example (actually one more idea for the test).

> I will have look at your implementation suggestion after :)
> 
> Edit: Regarding the error message: Reuse FS0855 or create a new error number etc. ?

The change I posted will currently reuse existing error (FS0017), which makes sense, I think, since it's technically not finding the correct override.

So, when we have something like:
```fsharp
type IOperation =
    static abstract member Execute: unit -> unit

type FaultyOperation() =
    interface IOperation with
        member _.Execute() = ()
```

it will result in

```log
error FS0017: The member 'Execute: unit -> unit' does not have the correct type to override the corresponding abstract method.
```

Here, when we emit it, we can do additional analysis to see what's not matching (if it's the isInstace flag, then we can produce different warning).

A slight problem might be when we have something like:
```fsharp
type IOperation =
    static abstract member Execute: unit -> unit

type FaultyOperation() =
    interface IOperation with
        member _.Execute() = ()
        static member Execute() = ()
```

It will still produce `FS0017` on the first member, which is _technically_, but probably better to produce `FS0855: No abstract or interface member was found that corresponds to this override`, which again, probably an additional check somewhere for the match of `isInstance`.

I can look at both if you'd like (probably tomorrow, slightly under the weather today).

**Comment by edgarfgp** [comment]
@vzarytovskii the Last commit uses your suggestion and adds more tests.

> I can look at both if you'd like (probably tomorrow, slightly under the weather today).

Hope you feel better tomorrow, Yes, please. I think we are close to having a fix for this. Thanks in advance

**Comment by vzarytovskii** [comment]
> @vzarytovskii the Last commit uses your suggestion and adds more tests.
> 
> 
> 
> > I can look at both if you'd like (probably tomorrow, slightly under the weather today).
> 
> 
> 
> Hope you feel better tomorrow, Yes, please. I think we are close to having a fix for this. Thanks in advance 
> 
> 
> 
> 

Cool, do you mind if I add to this PR?

**Comment by edgarfgp** [comment]
> Cool, do you mind if I add to this PR?

Not at all. I will be able to say that I worked with you in a fix ;)

**Comment by smoothdeveloper** [comment]
should the message be either:

> FS0855: No abstract or interface instance member was found that corresponds to this override

or 

> FS0855: No abstract or interface static member was found that corresponds to this override

**Comment by vzarytovskii** [comment]
> should the message be either:
> 
> 
> 
> > FS0855: No abstract or interface instance member was found that corresponds to this override
> 
> 
> 
> or 
> 
> 
> 
> > FS0855: No abstract or interface static member was found that corresponds to this override
> 
> 

That's why I mentioned why we can adjust it in place we throw/create it.

**Comment by vzarytovskii** [comment]
Fixed a bunch of stuff, I will probably need some help with figuring out tests, and gaps in the fix.

Now the behaviour is the following **(NOTE: all of those are from FSI, and not compiler, so will show only 1st error)**:

```fsharp
module StaticAbstractBug =
    type IOperation =
        static abstract member Execute: unit -> unit
        abstract member Execute2: unit -> unit
        static abstract member Property: int
        abstract member Property2: int
        static abstract Property3 : int with get, set

    type FaultyOperation() =
        interface IOperation with
            member _.Execute() = ()
            member _.Execute2() = ()
            member this.Property = 0
            member this.Property2 = 0
            member this.Property3 = 0
            member this.Property3 with set value = ()
```
will yield

```

              member _.Execute() = ()
  ---------------------^^^^^^^

/Users/u/code/fsharp2/src/stdin(16,22): error FS0855: No abstract or interface member was found that corresponds to this override
```
<hr/>
```fsharp
module StaticAbstractBug =
    type IOperation =
        static abstract member Execute: unit -> unit
        abstract member Execute: unit -> bool
        static abstract member Property: int
        abstract member Property: int

    type FaultyOperation() =
        interface IOperation with
            member _.Execute() = ()
            member _.Execute() = false
            member this.Property = 0
            member this.Property = false
```

will yield
```
             member _.Execute() = ()
  ---------------------------------^^
/Users/u/code/fsharp2/src/stdin(31,34): error FS0001: This expression was expected to have type
    'bool'    
but here has type
    'unit'   
```
**Need a sanity check, that this one is correct. It sees a non-static slot, which has `bool` as return type, but sees an implementation with unit.** 

<hr/>
```fsharp
module StaticAbstractBug =
    type IFoo<'T> =
       abstract DoIt: unit -> string
       static abstract Other : int -> int
       static abstract member Property: int
       abstract member Property2: int
       static abstract Property3 : int with get, set
    type MyFoo = {
       Value : int
    } with
      interface IFoo<MyFoo> with
        member me.DoIt() = string me.Value
        member _.Other(value) = value + 1
        member this.Property = 0
        member this.Property2 = 0
        member this.Property3 = 0
        member this.Property3 with set value = ()
```

yields
```
          member _.Other(value) = value + 1
  -----------------^^^^^

/Users/u/code/fsharp2/src/stdin(47,18): error FS0855: No abstract or interface member was found that corresponds to this override
```

I will check the rest and fix tests tomorrow.

**Comment by vzarytovskii** [comment]
Updated tests. Need some sanity check on issues.

File: src/Compiler/Checking/MethodOverrides.fs [hunk 4]
```diff
@@ -940,7 +947,7 @@ let GetAbstractMethInfosForSynMethodDecl(infoReader: InfoReader, ad, memberName:
             NameMultiMap.find  memberName.idText dispatchSlotsKeyed |> List.map (fun reqdSlot -> reqdSlot.MethodInfo)
         | ty, None -> 
             GetIntrinsicMethInfosOfType infoReader (Some memberName.idText) ad AllowMultiIntfInstantiations.Yes findFlag bindm ty
-    let dispatchSlots = minfos |> List.filter (fun minfo -> minfo.IsDispatchSlot)
+    let dispatchSlots = minfos |> List.filter (fun minfo -> minfo.IsDispatchSlot && minfo.IsInstance = memberFlags.IsInstance)
```
Comment (vzarytovskii) [inline src/Compiler/Checking/MethodOverrides.fs:950]: This pretty much filters out slots of different "instance-ness" (dunno more suitable word), so we have better error messages saying actually when something is not impmlemented/implemented but not there in the slots.

File: src/Compiler/Driver/CompilerDiagnostics.fs [hunk 5]
```diff
@@ -1547,6 +1549,14 @@ type Exception with
...
                         os.AppendString(OverrideDoesntOverride3E().Format sig2)
 
+                    // If implementation and required slot doesn't have same "instance-ness", then tell user that.
+                    if impl.IsInstance <> minfoVirt.IsInstance then
+                        // Requried slot is instance, meaning implementation is static, tell user that we expect instance.
+                        if minfoVirt.IsInstance then
+                            os.AppendString(OverrideShouldBeStatic().Format)
+                        else
+                            os.AppendString(OverrideShouldBeInstance().Format)
+
```
Comment (vzarytovskii) [inline src/Compiler/Driver/CompilerDiagnostics.fs:1559]: These will produce additional diagnostic message saying what was expected.

File: src/Compiler/Checking/CheckExpressions.fs [hunk 6]
```diff
@@ -11234,7 +11234,11 @@ and ApplyAbstractSlotInference (cenv: cenv) (envinner: TcEnv) (_: Val option) (a
              let uniqueAbstractMethSigs =
                  match dispatchSlots with
                  | [] ->
-                     errorR(Error(FSComp.SR.tcNoMemberFoundForOverride(), memberId.idRange))
+                     let instanceExpected = memberFlags.IsInstance
+                     if instanceExpected then
+                        errorR(Error(FSComp.SR.tcNoMemberFoundForOverride(), memberId.idRange))
+                     else
+                        errorR (Error(FSComp.SR.tcNoStaticMemberFoundForOverride (), memberId.idRange))
```
Comment (vzarytovskii) [inline src/Compiler/Checking/CheckExpressions.fs:11241]: These will now say whether it `static` member was expected or normal one (note, that I didn't change the message for normal one, since it would probably result in a metric tonne of baseline changes).

File: .vscode/launch.json [hunk 7]
```diff
@@ -21,7 +21,7 @@
             "type": "coreclr",
```
Comment (vzarytovskii) [inline .vscode/launch.json:21]: Added these, since was forgetting to rebuild and was feeling stupid that "nothing works", I say we leave them, since I wanter to add them to the tasks for a long time.

File: .vscode/tasks.json [hunk 8]
```diff
@@ -31,10 +31,18 @@
...
+                "env": {
+                    "UpdateXlfOnBuild": "true"
+                }
             },
             "problemMatcher": "$msCompile",
-            "group": "build",
+            "group": {
+                "kind": "build",
+                "isDefault": true
+            }
```
Comment (vzarytovskii) [inline .vscode/tasks.json:45]: Same here, meant to do that for a long time. But didn't have a PR for that. Default build just makes sense to have, and update xlfs on rebuild just makes sense.

**Comment by vzarytovskii** [comment]
@edgarfgp @smoothdeveloper if you have time, I'd really appreciate if you play with compiler/fsi with these changes to confirm I'm not crazy with my changes.

**Comment by edgarfgp** [comment]
> @edgarfgp @smoothdeveloper if you have time, I'd really appreciate if you play with compiler/fsi with these changes to confirm I'm not crazy with my changes.

Perfect. Will do

**Comment by vzarytovskii** [comment]
> > @edgarfgp @smoothdeveloper if you have time, I'd really appreciate if you play with compiler/fsi with these changes to confirm I'm not crazy with my changes.
> 
> Perfect. Will do

Thanks. I don't think I did anything with Properties though, so might be the gap.

The main change now is pretty much we distinguish not only name, arity and types, but also instance-ness.

**Comment by vzarytovskii** [comment]
tl;dr it's for sure not breaking and doesn't (shouldn't) introduce any new/changed messaging to not-static members, but will fix the original issue + give (imo) a bit more precise issues when we have mixed `static` and `instance` members in interface and implementation.

**Comment by edgarfgp** [comment]
> Thanks. I don't think I did anything with Properties, though, so that might be the gap.
> The main change now is pretty much we distinguish not only name, arity and types but also instance-ness.

@vzarytovskii  Updated and added more testing, and yeah we are not reporting `Properties ` both in positive and negative tests

Do you want to look into this, or should we create a separate issue to address this (Arguably our fix already fixed the linked issues)?

**Comment by vzarytovskii** [comment]
> not reporting Properties 

How can you give an example of what's expected and what's happening?

**Comment by edgarfgp** [comment]
> > not reporting Properties
> 
> How can you give an example of what's expected and what's happening?

#### Actual: Only reporting on instance members
#### Expected: Report on instance members and Properties

```fsharp
    [<Fact>]
    let ``Produce an error for interface with static abstract member that is implemented as instance member`` () =
        Fsx """
module StaticAbstractBug =
    type IFoo<'T> =
       abstract DoIt: unit -> string
       static abstract Other : int -> int
       static abstract member Property: int
       abstract member Property2: int
       static abstract Property3 : int with get, set
    type MyFoo = {
       Value : int
    } with
      interface IFoo<MyFoo> with
        member me.DoIt() = string me.Value
        member _.Other(value) = value + 1
        member this.Property = 0
        member this.Property2 = 0
        member this.Property3 = 0
        member this.Property3 with set value = ()
        """
         |> withOptions [ "--nowarn:3535" ]
         |> withLangVersion80
         |> compile
         |> shouldFail
         |> withSingleDiagnostic (Error 855, Line 14, Col 18, Line 14, Col 23, "No abstract or interface member was found that corresponds to this override")

```

### Actual: Only reporting on instance members
### Expected: Report on instance members and Properties

```fsharp
    [<Fact>]
    let ``Produces errors when includes keyword "static" when implementing a non generic interface in a type`` () =
        Fsx """
module StaticAbstractBug =
    type IFoo<'T> =
       abstract DoIt: unit -> string
       abstract Other : int -> int
       abstract member Property: int
       abstract member Property2: int
       abstract Property3 : int with get, set
    type MyFoo = {
       Value : int
    } with
      interface IFoo<MyFoo> with
        static member DoIt() = ""
        static member Other(value) = value + 1
        static member Property = 0
        static member Property2 = 0
        static member Property3 = 0
        static member Property3 with set value = ()
        """
         |> withLangVersion80
         |> typecheck
         |> shouldFail
         |> withDiagnostics [
             (Error 3855, Line 13, Col 23, Line 13, Col 27, "No static abstract member was found that corresponds to this override");
             (Error 3855, Line 14, Col 23, Line 14, Col 28, "No static abstract member was found that corresponds to this override")
         ]
```

**Comment by vzarytovskii** [comment]
So, the minimum repro would be:
Interface: have 1 static method, 1 static property
Implementation: have 1 instance method, 1 instance property

We expect both to be reported as missing/non-matching?

**Comment by edgarfgp** [comment]
Ok. I should easy based on the fix. Should be easy to extend and check for the properties using ie : 

3856,tcNoStaticPropertyFoundForOverride,"No static abstract property was found that corresponds to this override"

```fsharp
let instanceExpected = memberFlags.IsInstance
if instanceExpected then
    errorR(Error(FSComp.SR.tcNoPropertyFoundForOverride(), memberId.idRange))
else
    errorR (Error(FSComp.SR.tcNoStaticPropertyFoundForOverride (), memberId.idRange))     
```

**Comment by T-Gro** [comment]
Before this gets merged, @edgarfgp  could you please also rename names used in the code (e.g. keys used in FsComp.txt) to call them `implements` and not `overrides`?

**Comment by edgarfgp** [comment]
> Before this gets merged, @edgarfgp could you please also rename names used in the code (e.g. keys used in FsComp.txt) to call them `implements` and not `overrides`?

Sure

**Comment by vzarytovskii** [comment]
> Ok. I should easy based on the fix. Should be easy to extend and check for the properties using ie : 
> 
> 
> 
> 3856,tcNoStaticPropertyFoundForOverride,"No static abstract property was found that corresponds to this override"
> 
> 
> 
> ```fsharp
> 
> let instanceExpected = memberFlags.IsInstance
> 
> if instanceExpected then
> 
>     errorR(Error(FSComp.SR.tcNoPropertyFoundForOverride(), memberId.idRange))
> 
> else
> 
>     errorR (Error(FSComp.SR.tcNoStaticPropertyFoundForOverride (), memberId.idRange))     
> 
> ```
> 
> 

It's not only about adding message. It needs a change in logic when searching for properties' slots.

**Comment by edgarfgp** [comment]
> It's not only about adding message. It needs a change in logic when searching for properties' slots.

Yeah I figured after giving is try 😅.

**Comment by vzarytovskii** [comment]
> > It's not only about adding message. It needs a change in logic when searching for properties' slots.
> 
> 
> 
> Yeah I figured after giving is try 😅. 

I'll fix it. Is my previous comment true? Is it only thing we should do?

**Comment by edgarfgp** [comment]
> We expect both to be reported as missing/non-matching?

Sorry I missed this . Yea that would be ideal to report in both scenarios

**Comment by vzarytovskii** [comment]
I will get back to it tomorrow, if that's ok.

**Review by vzarytovskii — APPROVED** [review]
Thanks!
